# David Lennox-Hvenekilde 22/10/20
# v2.0
# This script takes a UnniProt table of BioB sequences, with various sequence and phylogeny, IDs and clean it up
# This uses the final version of the motif (v4) for data analysis, for previosu version and the iterative improvement done to the 
# Motif, see v1.0 "Cleanup_allBioB_Uniprot.R"
# Databases used:
# UniProt release 2020_05
# Search string "family:biotin family:synthase"

# C-x(3)-C-x(2)-C-x(25,45)-[HSRGEAKT]-[FYVLI]-C/S/G/A/D-[FILMV]-[VAG]-[ATSVLIM]-[ASGQ]-[WGEQY]-[KRTYELSAM]

#==========================================================================================================
# Load libraries and data#################################################################################
library(stringr)
library(plyr)
library(ggplot2)
library(seqinr)
library(dplyr)
library(tidyr)
library(RColorBrewer)
library(data.table)
library(ggsci)
## Needed for extra fonts
#(extrafont)
#font_import()
#loadfonts(device = "win")


rm(list = ls(all.names = TRUE))
dir_name <-dirname(rstudioapi::getSourceEditorContext()$path)
setwd(dir_name)
#Read and clean up the main UniProt BioB database file
BioB.data<-read.delim("201020 ALL UniProt BioBs with phylogeny")
BioB.data$Organism_Clean <- word(BioB.data$Organism, 1,2, sep=" ")

# Read in the motif data, generated by ScanProsite, using the five motifs indicated above (C-TypeI/S-Type II/G/A/D)
# Against the downloaded UniProt BioB file
BioB_motifv4<-read.delim("./V4 motifs/201022 V4 motifs.txt")
BioB_motifv4$Entry<-word(BioB_motifv4$ID,1, sep="\\|")

# Join the two files
BioB.data.motifv4<-join(BioB.data, BioB_motifv4, by="Entry")
# How many have double motifs per sequence? 138
nrow(BioB.data.motifv4)-nrow(BioB.data)
# How many have no motif? 506
nrow(subset(BioB.data.motifv4,is.na(BioB.data.motifv4$Motif)))
# lets look more closely at the 138 sequences with more than one motif:
Motifs_countsv4 <- data.frame(plyr::count(BioB.data.motifv4$Entry))
Motifs_countsv4 <- subset(Motifs_countsv4,Motifs_countsv4$freq>1)
Motifs_countsv4.fulldata<-subset(BioB.data.motifv4,BioB.data.motifv4$Entry==Motifs_countsv4$x)
Motifs_countsv4.fulldata<-BioB.data.motifv4[BioB.data.motifv4$Entry %in% Motifs_countsv4$x,]
#======================================================================================================================
# Cleaning up multiple motifs per sequence#############################################################################
# Looking at the data it seems that many of the motifs are "anchored to the same CxxxCxxC SAM radical motif
# This means that they are most likely false positives. It looks like the first motif is the correct one, and that
# we allowed for too high distance between the SAM radical and Auxillary cluster in the sequence. We therefore 
# remove the second, false motif. However if the motifs dont anchor at the same CxxxCxxC they are both kept
NoDup<-Motifs_countsv4.fulldata[1,]
NoDup<-NoDup[-1,]

for (EntryID in unique(Motifs_countsv4.fulldata$Entry)){
  # Fetch the rows with the unique ID, there are two
  db1 <- subset(Motifs_countsv4.fulldata, Motifs_countsv4.fulldata$Entry==EntryID)
  #If there the start of the motifs is the same (anchored to the same SAM-radical motif)
  # Then we choose the first of the occurring motifs as "real"
  if(db1$Start[1]==db1$Start[2]){
    db2<-db1[sapply(db1$End, which.min),][1,]
    NoDup<-rbind(NoDup,db2)
    
    #However, if the CxxxCxxC anchor for the two motifs in the same sequence is NOT the same
    #Then we assume that this is not a false positive 
  } else{
    NoDup<-rbind(NoDup,db1)
  }
}

# After filtering out the ones same anchor CxxxCxxC, what sequences with two motifs we have left
NoDup_2ormore <- data.frame(plyr::count(NoDup$Entry))
NoDup_2ormore <- subset(NoDup_2ormore,NoDup_2ormore$freq>1)
NoDup_2ormore.fulldata<-BioB.data.motifv4[BioB.data.motifv4$Entry %in% NoDup_2ormore$x,]
#write.csv(NoDup_2ormore.fulldata, "Doulbe_motif_clean.csv")
# How many streptomyces sequences are there?

# It seems that it is only Streptomyces and one Pythium. The motifs are separated by a very wide margin. 
# Now we generate the final data-frame on which we wil work:
Motifs4.dubiltered.fulldata<-BioB.data.motifv4[!(BioB.data.motifv4$Entry %in% Motifs_countsv4$x),]
BioB_motifs_clean<-rbind(Motifs4.dubiltered.fulldata,NoDup)
#write.csv(BioB_motifs_clean, "BioB_motifs_clean.csv")

NoMotif<-subset(BioB.data.motifv4,is.na(BioB.data.motifv4$Motif))
#Rando100<-sample_n(NoMotif,100)
#write.fasta(as.list(Rando100$Sequence), names=Rando100$Entry, 
#            as.string=FALSE, file.out="100RandomNoMotifv4.fa")

# Looking at the sequence alignment of the random 100 biobs outside motifv4, it was decided not to refine the motif any further
506/23268*100
# 2.17% of the sequences have no motif. This is OK. There might be some cases where the sequences are imcomplete, or misannotated
# Inspection also revealed that some Archea sequences don't have the CxxxCxxC motif, but rather a CxxxxxCxxC
# for example: A0A2I0PIW6, T2GJS2, R7PV60

#======================================================================================================================
# Lets look at the distributions of motifs ############################################################################
rm(list = ls(all.names = TRUE))
BioB_motifs_clean <- read.csv("BioB_motifs_clean.csv")
BioB_motifs_clean_reduced = BioB_motifs_clean
#BioB_motifs_clean[!duplicated(BioB_motifs_clean$Organism_Clean),]
length(unique(subset(BioB_motifs_clean,BioB_motifs_clean$Taxonomic.lineage..GENUS.=="Streptomyces")$Entry.name))
# Eukaryotes only encode type I BioBs
BioB_motifs_clean_euk<-BioB_motifs_clean[BioB_motifs_clean$Taxonomic.lineage..SUPERKINGDOM.=="Eukaryota",]
unique(BioB_motifs_clean_euk$Motif)
# Archaea
BioB_motifs_clean_arc<-BioB_motifs_clean[BioB_motifs_clean$Taxonomic.lineage..SUPERKINGDOM.=="Archaea",]
unique(BioB_motifs_clean_arc$Motif)


# Export All streptomyces sequences
BioB_motifs_clean_strep <- subset(BioB_motifs_clean,BioB_motifs_clean$Taxonomic.lineage..GENUS.=="Streptomyces")
BioB_motifs_clean_strep2 <- BioB_motifs_clean_strep[!duplicated(BioB_motifs_clean_strep$Entry),]
# Write all strep biobs
write.fasta(as.list(BioB_motifs_clean_strep2$Sequence), 
            names=BioB_motifs_clean_strep2$Entry, 
            as.string=FALSE, file.out="StreptomycesBioBs.fa")
# Write only strep sequences with two motifs
BioB_motifs_clean_strep_dub <- BioB_motifs_clean_strep[duplicated(BioB_motifs_clean_strep$Entry),]
write.fasta(as.list(BioB_motifs_clean_strep_dub$Sequence), 
            names=BioB_motifs_clean_strep_dub$Entry, 
            as.string=FALSE, file.out="StreptomycesBioBs_dub.fa")



Phylum.counts <- data.frame(plyr::count(BioB_motifs_clean_reduced$Taxonomic.lineage..PHYLUM.))
# Then add the counts for type I motif BioB
Phylum.counts <- join(Phylum.counts,data.frame(
  plyr::count(subset(BioB_motifs_clean_reduced,BioB_motifs_clean_reduced$Motif == "TI")$Taxonomic.lineage..PHYLUM.)), by="x")
# Then add the counts for type II motif BioB
Phylum.counts <- join(Phylum.counts,data.frame(
  plyr::count(subset(BioB_motifs_clean_reduced,BioB_motifs_clean_reduced$Motif == "TII")$Taxonomic.lineage..PHYLUM.)), by="x")

Phylum.counts <- join(Phylum.counts,data.frame(
  plyr::count(subset(BioB_motifs_clean_reduced,BioB_motifs_clean_reduced$Motif == "TG")$Taxonomic.lineage..PHYLUM.)), by="x")

Phylum.counts <- join(Phylum.counts,data.frame(
  plyr::count(subset(BioB_motifs_clean_reduced,BioB_motifs_clean_reduced$Motif == "TA")$Taxonomic.lineage..PHYLUM.)), by="x")

Phylum.counts <- join(Phylum.counts,data.frame(
  plyr::count(subset(BioB_motifs_clean_reduced,BioB_motifs_clean_reduced$Motif == "TD")$Taxonomic.lineage..PHYLUM.)), by="x")

Phylum.counts <- join(Phylum.counts,data.frame(
  plyr::count(subset(BioB_motifs_clean_reduced,is.na(BioB_motifs_clean_reduced$Motif))$Taxonomic.lineage..PHYLUM.)), by="x")

colnames(Phylum.counts)<-c("Phylum","Total", "TI", "TII", "TG", "TA", "TD", "NA")
Phylum.counts[is.na(Phylum.counts)] <- 0

# Plot the phylum and the counts of TI and TII
# First the DF needs to long-form to work in ggplot2
PhylumDF <- gather(Phylum.counts, Motif, Per_Motif, TI:"NA")

# We now add the Kingdom
PhylumDF$Kingdom<-NA
for (phy in unique(PhylumDF$Phylum)){
  PhylumDF[PhylumDF$Phylum==phy,]$Kingdom<-BioB_motifs_clean_reduced[BioB_motifs_clean_reduced$Taxonomic.lineage..PHYLUM.==phy,][1,]$Taxonomic.lineage..SUPERKINGDOM.
}

# Lets reduce it to the bigger phylum groups
# First remove the single virus sequence
PhylumDF<-PhylumDF[PhylumDF$Kingdom!="Viruses",]
PhylumDF[PhylumDF$Phylum=="",]$Phylum<-"No Phylum"
PhylumDF_reduced<-PhylumDF[PhylumDF$Total>20,]

# Kingdom plot ########################################################################################
p3<-ggplot(PhylumDF, aes(fill=Motif, y=Per_Motif, x=Kingdom)) + 
  geom_bar(position="fill", stat="identity") +
  labs(title="Distribution of BioB auxillary cluster \ncoordination motifs across Kingdoms", y="Percent")+
  theme(axis.title.x = element_blank(),
        text = element_text(size = 18, family = "Garamond", color = "grey20")
        )+
  scale_fill_npg()
p3
         #family="Comic Sans MS"))
# Export the plot
image=ggplot(PhylumDF, aes(fill=Motif, y=Per_Motif, x=Kingdom)) + 
  geom_bar(position="fill", stat="identity") +
  labs(title="Distribution of BioB auxillary cluster \ncoordination motifs across Kingdoms", y="Percent")+
  theme(axis.title.x = element_blank(),
        text = element_text(size = 18, family = "Garamond", color = "grey20")
  )+
  scale_fill_npg()
ggsave(file="./Exported images/Kingdom_motif_distributionV2.svg", plot=image, width=10, height=8)

# How many are there of each Kingdom??
c(sum(PhylumDF[PhylumDF$Kingdom=="Archaea",]$Per_Motif),sum(PhylumDF[PhylumDF$Kingdom=="Bacteria",]$Per_Motif),sum(PhylumDF[PhylumDF$Kingdom=="Eukaryota",]$Per_Motif))
c(nrow(BioB.data[BioB.data$Taxonomic.lineage..SUPERKINGDOM.=="Archaea",]),nrow(BioB.data[BioB.data$Taxonomic.lineage..SUPERKINGDOM.=="Bacteria",]),nrow(BioB.data[BioB.data$Taxonomic.lineage..SUPERKINGDOM.=="Eukaryota",]))

# Bacterial phylum plot ########################################################################################
PhylumDF_reduced_bacteria<-PhylumDF_reduced[PhylumDF_reduced$Kingdom=="Bacteria",]
p4<-ggplot(PhylumDF_reduced_bacteria, aes(fill=Motif, y=Per_Motif, x=Phylum)) + 
  geom_bar(position="fill", stat="identity") +
  labs(title="Distribution of BioB auxillary cluster \ncoordination motifs across most abundant Bacterial Phyla", y="Percent")+
  theme(axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        legend.position = "none",
        text = element_text(size = 18, family = "Garamond", color = "grey20"),
        axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1))+
  scale_fill_npg()
p4
# Export the plot
# image=p4
# ggsave(file="./Exported images/Bacteria_Phyla_motif_distribution.svg", plot=image, width=10, height=8)

#Make a combination plot with both Kingdoms and Phyla
require(gridExtra)
plot1 <-p3
plot2 <-p4
grid.arrange(plot1, plot2, ncol=2)
ggsave(file="./Exported images/combinationPlotV1.svg", plot=grid.arrange(plot1, plot2, ncol=2), width=16, height=8)



# Lets have a look ath the phyulum firmicutes ################################################################
Firm<-BioB_motifs_clean[BioB_motifs_clean$Taxonomic.lineage..PHYLUM.=="Firmicutes",]
Firm.order.counts <- data.frame(plyr::count(Firm$Taxonomic.lineage..ORDER.))
# Then add the counts for type I motif BioB
Firm.order.counts <- join(Firm.order.counts,data.frame(
  plyr::count(subset(Firm,Firm$Motif == "TI")$Taxonomic.lineage..ORDER.)), by="x")
# Then add the counts for type II motif BioB
Firm.order.counts <- join(Firm.order.counts,data.frame(
  plyr::count(subset(Firm,Firm$Motif == "TII")$Taxonomic.lineage..ORDER.)), by="x")

Firm.order.counts <- join(Firm.order.counts,data.frame(
  plyr::count(subset(Firm,Firm$Motif == "TG")$Taxonomic.lineage..ORDER.)), by="x")

Firm.order.counts <- join(Firm.order.counts,data.frame(
  plyr::count(subset(Firm,Firm$Motif == "TA")$Taxonomic.lineage..ORDER.)), by="x")

Firm.order.counts <- join(Firm.order.counts,data.frame(
  plyr::count(subset(Firm,Firm$Motif == "TD")$Taxonomic.lineage..ORDER.)), by="x")

Firm.order.counts <- join(Firm.order.counts,data.frame(
  plyr::count(subset(Firm,is.na(Firm$Motif))$Taxonomic.lineage..ORDER.)), by="x")

colnames(Firm.order.counts)<-c("Order","Total", "TI", "TII", "TG", "TA", "TD", "NA")
Firm.order.counts[is.na(Firm.order.counts)] <- 0

Firm.order.counts.long <- gather(Firm.order.counts, Motif, Per_Motif, TI:"NA")

p5<-ggplot(Firm.order.counts.long, aes(fill=Motif, y=Per_Motif, x=Order)) + 
  geom_bar(position="fill", stat="identity") +
  labs(title="Distribution of BioB auxillary cluster \ncoordination motifs Firmicutes Phylum", y="Percent")+
  theme(axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        text = element_text(size = 18, family = "Arial", color = "grey20"),
        axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1))+
  scale_fill_npg()
p5
ggsave(file="./Exported images/Firmicutes_Order_Motif.svg", plot=p5, width=10, height=8)

############################################################################################################################
# Since it seem that alternative coordination is quite normal in obligate anaerobes, lets have a closer look
# To do this, lets link some META-data on e.g. oxygen tolerance of the microorganisms, our "database" of Biotin Synthase sequences
library(RCurl)
library(rjson)
formatjson <- '/?format=json'
page <- '1'
json <- '&format=json'
taxon <- 'https://bacdive.dsmz.de/api/pnu/taxon/'
searchterm <- 'acetobacter'
url_taxon <- URLencode(paste0(taxon,searchterm,formatjson))

response <- getURL(url_taxon,userpwd="{dal@biosyntia.com}:{Moz85314b}", httpauth = 1L)



############################################################################################################################
# EXPORT OF DATA FOR PHYLOGENETIC TREE
set.seed(12345)
# We remove species duplication to account of overweight of a few species (e.g. Escherichia coli)
BioB_motifs_clean_reduced = BioB_motifs_clean[!duplicated(BioB_motifs_clean$Organism_Clean),]
# How stil it up in 100 random
BioB_motifs_clean_reduced_TI <- subset(BioB_motifs_clean_reduced, BioB_motifs_clean_reduced$Motif =="TI")
TI_rando<-BioB_motifs_clean_reduced_TI[sample(nrow(BioB_motifs_clean_reduced_TI), 500), ]
# How stil it up in 100 random
BioB_motifs_clean_reduced_TII <- subset(BioB_motifs_clean_reduced, BioB_motifs_clean_reduced$Motif =="TII")
TII_rando<-BioB_motifs_clean_reduced_TII[sample(nrow(BioB_motifs_clean_reduced_TII), 500), ]
# How stil it up in 100 random
BioB_motifs_clean_reduced_TG <- subset(BioB_motifs_clean_reduced, BioB_motifs_clean_reduced$Motif =="TG")
TG_rando<-BioB_motifs_clean_reduced_TG[sample(nrow(BioB_motifs_clean_reduced_TG), 50), ]
# How stil it up in 100 random
BioB_motifs_clean_reduced_TA <- subset(BioB_motifs_clean_reduced, BioB_motifs_clean_reduced$Motif =="TA")
TA_rando<-BioB_motifs_clean_reduced_TA[sample(nrow(BioB_motifs_clean_reduced_TA), 50), ]
# How stil it up in 100 random
BioB_motifs_clean_reduced_TD <- subset(BioB_motifs_clean_reduced, BioB_motifs_clean_reduced$Motif =="TD")
TD_rando<-BioB_motifs_clean_reduced_TD[sample(nrow(BioB_motifs_clean_reduced_TD), 50), ]


ExportTI_TII<-rbind(TI_rando, TII_rando)
#write.csv(ExportTI_TII, "201028_ExportTI_TII.csv")

#write.fasta(as.list(ExportTI_TII$Sequence), names=ExportTI_TII$Entry, 
#           as.string=FALSE, file.out="201028ExportTI_TII.fa")



# Export nr 2. Top bacterial genera
cout_phylum<-BioB_motifs_clean_reduced[BioB_motifs_clean_reduced$Taxonomic.lineage..SUPERKINGDOM.=="Bacteria",]
cout_phylum<-data.frame(plyr::count(cout_phylum$Taxonomic.lineage..GENUS.))
cout_phylum<-cout_phylum[order(cout_phylum[,2],decreasing = TRUE),]
cout_phylum$x[1:40]

BioB_motifs_clean_reduced_topGenera<-subset(BioB_motifs_clean_reduced, BioB_motifs_clean_reduced$Taxonomic.lineage..GENUS %in% cout_phylum$x[1:40])
# Split up well known genera by oxygen tolerance
Anaerobic<- c("Bacteroides", "Clostridium", "Fusobacterium", "Veillonella", "Porphyromonas", "Ruminococcus", 
              "Blautia", "Desulfotomaculum", "Brachyspira", "Desulfovibrio", "Caldicellulosiruptor", "Roseburia" )

Aerobic<-c("Pseudomonas","Bacillus", "Streptomyces", "Flavobacterium","Mycobacterium", "Leptospira", "Acinetobacter", 
           "Legionella", "Burkholderia","Sphingomonas", "Micromonospora", "Xanthomonas"  )

Facultative<-c("Corynebacterium", "Vibrio", "Chryseobacterium", "Staphylococcus", "Escherichia", "Campylobacter", 
               "Salmonella", "Helicobacter", "Yersinia", "Paracoccus", "Serratia", "Shewanella" )

BioBs_genus_oxygen_subset<-subset(BioB_motifs_clean_reduced, BioB_motifs_clean_reduced$Taxonomic.lineage..GENUS %in% c(Anaerobic, Aerobic, Facultative))
BioBs_genus_oxygen_subset$Oxygen<-"na"
BioBs_genus_oxygen_subset[BioBs_genus_oxygen_subset$Taxonomic.lineage..GENUS %in% Anaerobic,]$Oxygen<-"Anaerobic"
BioBs_genus_oxygen_subset[BioBs_genus_oxygen_subset$Taxonomic.lineage..GENUS %in% Aerobic,]$Oxygen<-"Aerobic"
BioBs_genus_oxygen_subset[BioBs_genus_oxygen_subset$Taxonomic.lineage..GENUS %in% Facultative,]$Oxygen<-"Facultative"

#Lets take 10 random sequences from each of the 30 genera that we have chose, so none of them are too over represented
df_oxy_genera <- data.frame(matrix(ncol = ncol(BioBs_genus_oxygen_subset), nrow = 0))
x <- colnames(BioBs_genus_oxygen_subset)
colnames(df_oxy_genera) <- x

# For each unique genus, we take 5 random sequences and add them to the empty DF just made
for (gen in unique(BioBs_genus_oxygen_subset$Taxonomic.lineage..GENUS.)){
  
  BioBs_genus_oxygen_subset_genus<-subset(BioBs_genus_oxygen_subset, BioBs_genus_oxygen_subset$Taxonomic.lineage..GENUS.==gen)
  
  df_oxy_genera <- rbind(df_oxy_genera,
                         BioBs_genus_oxygen_subset_genus[sample(nrow(BioBs_genus_oxygen_subset_genus), 5), ] )
}


# Include the colors for annotation in iTOL###################################
df_oxy_genera$Motif_col<-"nada"

NA_col<-"#E64B35"
df_oxy_genera[is.na(df_oxy_genera$Motif), ]$Motif_col<-NA_col
TI_col<-"#F39B7F"
df_oxy_genera[which(df_oxy_genera$Motif=="TI"), ]$Motif_col<-TI_col
TII_col<-"#8491B4"
df_oxy_genera[which(df_oxy_genera$Motif=="TII"), ]$Motif_col<-TII_col
TG_col<-"#3C5488"
df_oxy_genera[which(df_oxy_genera$Motif=="TG"), ]$Motif_col<-TG_col
TD_col<-"#00A087"
df_oxy_genera[which(df_oxy_genera$Motif=="TD"), ]$Motif_col<-TD_col
TA_col<-"#4DBBD5"
df_oxy_genera[which(df_oxy_genera$Motif=="TA"), ]$Motif_col<-TA_col


df_oxy_genera$Oxy_col<-"nada"
Anaerobic_col<-"#4a4a4a"
df_oxy_genera[which(df_oxy_genera$Oxygen=="Anaerobic"), ]$Oxy_col<-Anaerobic_col

Facultative_col<-"#808080"
df_oxy_genera[which(df_oxy_genera$Oxygen=="Facultative"), ]$Oxy_col<-Facultative_col

Aerobic_col<-"#cccccc"
df_oxy_genera[which(df_oxy_genera$Oxygen=="Aerobic"), ]$Oxy_col<-Aerobic_col

df_oxy_genera$Fullname<-paste(df_oxy_genera$Organism_Clean,"-",df_oxy_genera$Entry)


write.csv(df_oxy_genera, "df_oxy_genera_color.csv")

write.fasta(as.list(df_oxy_genera$Sequence), names=df_oxy_genera$Entry, 
             as.string=FALSE, file.out="Oxygen_genus_subset.fa")

# Test the length of the different BioB types###################################################################
BioB_motifs_clean_reduced_bac<-subset(BioB_motifs_clean_reduced,BioB_motifs_clean_reduced$Taxonomic.lineage..SUPERKINGDOM.=="Bacteria") 

length(unique(BioB_motifs_clean_reduced_bac$Taxonomic.lineage..PHYLUM.))


p6<- ggplot(BioB_motifs_clean_reduced_bac, aes(x=Motif, y=Length)) + 
  geom_boxplot()+
  labs(title="Bacterial BioB sequence length \n across ", y="Percent")+
  theme(axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        text = element_text(size = 18, family = "Arial", color = "grey20"),
        axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1))+
  scale_fill_npg()

p6
p6+geom_jitter(16, position=position_jitter(0.2))

ggplot(BioB_motifs_clean_reduced_bac, aes(x=Motif, y=Length, color=Taxonomic.lineage..PHYLUM.)) + 
  geom_jitter(position=position_jitter(0.2), cex=0.2)

str_split_fixed(BioB_motifs_clean_reduced_bac$Taxonomic.lineage..PHYLUM., pattern, n)
BioB_motifs_clean_reduced_bac_Nocand<-subset(BioB_motifs_clean_reduced_bac, word(BioB_motifs_clean_reduced_bac$Taxonomic.lineage..PHYLUM., 1) != c("Candidatus"))
BioB_motifs_clean_reduced_bac_Nocand<-subset(BioB_motifs_clean_reduced_bac_Nocand, word(BioB_motifs_clean_reduced_bac_Nocand$Taxonomic.lineage..PHYLUM., 1) != c("candidate"))

ggplot(BioB_motifs_clean_reduced_bac_Nocand, aes(x=Motif, y=Length, color=Taxonomic.lineage..PHYLUM.)) + 
  geom_jitter(position=position_jitter(0.2), cex=0.2)

##############################################################################################################
# Looking at the Cyanbacterial BioBs
CyanoBioB<-subset(BioB_motifs_clean_reduced, BioB_motifs_clean_reduced$Taxonomic.lineage..PHYLUM.=="Cyanobacteria")
# These are all annotated with Type I
# Lets compare them with some of the previosuly discusses Type I and Type II BioBs from anaerobes or aerobes
BioBs_genus_oxygen_subset_anaerobic<-BioBs_genus_oxygen_subset[BioBs_genus_oxygen_subset$Oxygen=="Anaerobic",]
BioBs_genus_oxygen_subset_aerobic<-BioBs_genus_oxygen_subset[BioBs_genus_oxygen_subset$Oxygen=="Aerobic",]
BioBs_genus_oxygen_subset_facultative<-BioBs_genus_oxygen_subset[BioBs_genus_oxygen_subset$Oxygen=="Facultative",]

Anaerobic_20_rando<-BioBs_genus_oxygen_subset_anaerobic[sample(nrow(BioBs_genus_oxygen_subset_anaerobic), 20), ]
Aerobic_20_rando<-BioBs_genus_oxygen_subset_aerobic[sample(nrow(BioBs_genus_oxygen_subset_aerobic), 20), ]
Facultative_20_rando<-BioBs_genus_oxygen_subset_facultative[sample(nrow(BioBs_genus_oxygen_subset_facultative), 20), ]

CyanoBioB$Oxygen <- "Photosynthetic"
CyanoAndMore_dataset<-rbind(CyanoBioB, Anaerobic_20_rando, Aerobic_20_rando, Facultative_20_rando)

# Color for iTOL
CyanoAndMore_dataset$Oxy_col<-"nada"
Anaerobic_col<-"#4a4a4a"
CyanoAndMore_dataset[which(CyanoAndMore_dataset$Oxygen=="Anaerobic"), ]$Oxy_col<-Anaerobic_col

Facultative_col<-"#808080"
CyanoAndMore_dataset[which(CyanoAndMore_dataset$Oxygen=="Facultative"), ]$Oxy_col<-Facultative_col

Aerobic_col<-"#cccccc"
CyanoAndMore_dataset[which(CyanoAndMore_dataset$Oxygen=="Aerobic"), ]$Oxy_col<-Aerobic_col

Photosynthetic_col<-"#000000"
CyanoAndMore_dataset[which(CyanoAndMore_dataset$Oxygen=="Photosynthetic"), ]$Oxy_col<-Photosynthetic_col

CyanoAndMore_dataset$Fullname<-paste(CyanoAndMore_dataset$Organism_Clean,"-",CyanoAndMore_dataset$Entry)
# 
CyanoAndMore_dataset$Motif_col<-"nada"

# Color my Motif type ######################################################################
NA_col<-"#E64B35"
CyanoAndMore_dataset[is.na(CyanoAndMore_dataset$Motif), ]$Motif_col<-NA_col
TI_col<-"#F39B7F"
CyanoAndMore_dataset[which(CyanoAndMore_dataset$Motif=="TI"), ]$Motif_col<-TI_col
TII_col<-"#8491B4"
CyanoAndMore_dataset[which(CyanoAndMore_dataset$Motif=="TII"), ]$Motif_col<-TII_col
TG_col<-"#3C5488"
CyanoAndMore_dataset[which(CyanoAndMore_dataset$Motif=="TG"), ]$Motif_col<-TG_col
TD_col<-"#00A087"
CyanoAndMore_dataset[which(CyanoAndMore_dataset$Motif=="TD"), ]$Motif_col<-TD_col
TA_col<-"#4DBBD5"
CyanoAndMore_dataset[which(CyanoAndMore_dataset$Motif=="TA"), ]$Motif_col<-TA_col

# Export ######################################################################
write.csv(CyanoAndMore_dataset, "CyanoAndMore_dataset.csv")
nrow(CyanoAndMore_dataset)
length(CyanoAndMore_dataset$Sequence)
length(CyanoAndMore_dataset$Entry)
write.fasta(as.list(CyanoAndMore_dataset$Sequence), names=CyanoAndMore_dataset$Entry, 
            as.string=FALSE, file.out="Cyano_subset.fa")

# Generating fasta files for whole motif alignments ##############################
#Load the data, make sure BioB_motifs_clean.csv is present in the same dir as this script
rm(list = ls(all.names = TRUE))
dir_name <-dirname(rstudioapi::getSourceEditorContext()$path)
setwd(dir_name)
BioB_motifs_clean <- read.csv("BioB_motifs_clean.csv")

# We extract E coli K-12 BioB as the "control" sequence which we will base the alignments on
EcoliK12<-subset(BioB_motifs_clean,BioB_motifs_clean$Entry.name=="BIOB_ECOLI")

nrow(subset(BioB_motifs_clean,BioB_motifs_clean$Motif=="TI"))
# There are 20254 TI BioBs, Clustal Omega only takes 4000 seq at a time. They will be aligned in 6 rounds
cutoff<-as.integer(nrow(subset(BioB_motifs_clean,BioB_motifs_clean$Motif=="TI"))/6)

#Writing 6 fasta files with 1/6 of the TI sequences each, for manual run in EMBL-EBI CLUSTAL-OMEGA:
#All 6 files contain the E. coli K-12 biob sequence as a reference sequence
# 1
write.fasta(as.list(subset(BioB_motifs_clean,BioB_motifs_clean$Motif=="TI")[1:cutoff,]$Sequence), 
            names=subset(BioB_motifs_clean,BioB_motifs_clean$Motif=="TI")[1:cutoff,]$Entry, 
            as.string=FALSE, file.out="TypeI_tablexport_1.fa")

# 2
write.fasta(as.list(c(EcoliK12$Sequence,subset(BioB_motifs_clean,BioB_motifs_clean$Motif=="TI")[(cutoff+1):(2*cutoff),]$Sequence)), 
            names=c(EcoliK12$Entry,subset(BioB_motifs_clean,BioB_motifs_clean$Motif=="TI")[(cutoff+1):(2*cutoff),]$Entry), 
            as.string=FALSE, file.out="TypeI_tablexport_2.fa")

# 3
write.fasta(as.list(c(EcoliK12$Sequence,subset(BioB_motifs_clean,BioB_motifs_clean$Motif=="TI")[(2*cutoff+1):(3*cutoff),]$Sequence)), 
            names=c(EcoliK12$Entry,subset(BioB_motifs_clean,BioB_motifs_clean$Motif=="TI")[(2*cutoff+1):(3*cutoff),]$Entry), 
            as.string=FALSE, file.out="TypeI_tablexport_3.fa")

# 4
write.fasta(as.list(c(EcoliK12$Sequence,subset(BioB_motifs_clean,BioB_motifs_clean$Motif=="TI")[(3*cutoff+1):(4*cutoff),]$Sequence)), 
            names=c(EcoliK12$Entry,subset(BioB_motifs_clean,BioB_motifs_clean$Motif=="TI")[(3*cutoff+1):(4*cutoff),]$Entry), 
            as.string=FALSE, file.out="TypeI_tablexport_4.fa")

# 5
write.fasta(as.list(c(EcoliK12$Sequence,subset(BioB_motifs_clean,BioB_motifs_clean$Motif=="TI")[(4*cutoff+1):(5*cutoff),]$Sequence)), 
            names=c(EcoliK12$Entry,subset(BioB_motifs_clean,BioB_motifs_clean$Motif=="TI")[(4*cutoff+1):(5*cutoff),]$Entry), 
            as.string=FALSE, file.out="TypeI_tablexport_5.fa")

# 6
write.fasta(as.list(c(EcoliK12$Sequence,subset(BioB_motifs_clean,BioB_motifs_clean$Motif=="TI")[(5*cutoff+1):nrow(subset(BioB_motifs_clean,BioB_motifs_clean$Motif=="TI")),]$Sequence)), 
            names=c(EcoliK12$Entry,subset(BioB_motifs_clean,BioB_motifs_clean$Motif=="TI")[(5*cutoff+1):nrow(subset(BioB_motifs_clean,BioB_motifs_clean$Motif=="TI")),]$Entry), 
            as.string=FALSE, file.out="TypeI_tablexport_6.fa")


# Do the same with Type II, don't need several files.
unique(BioB_motifs_clean$Motif)

nrow(subset(BioB_motifs_clean,BioB_motifs_clean$Motif=="TII"))
write.fasta(as.list(c(EcoliK12$Sequence,subset(BioB_motifs_clean,BioB_motifs_clean$Motif=="TII")$Sequence)), 
            names=c(EcoliK12$Entry,subset(BioB_motifs_clean,BioB_motifs_clean$Motif=="TII")$Entry), 
            as.string=FALSE, file.out="TypeII_tablexport.fa")

write.fasta(as.list(c(EcoliK12$Sequence,subset(BioB_motifs_clean,BioB_motifs_clean$Motif=="TA")$Sequence)), 
            names=c(EcoliK12$Entry,subset(BioB_motifs_clean,BioB_motifs_clean$Motif=="TA")$Entry), 
            as.string=FALSE, file.out="TypeTA_tablexport.fa")

write.fasta(as.list(c(EcoliK12$Sequence,subset(BioB_motifs_clean,BioB_motifs_clean$Motif=="TD")$Sequence)), 
            names=c(EcoliK12$Entry,subset(BioB_motifs_clean,BioB_motifs_clean$Motif=="TD")$Entry), 
            as.string=FALSE, file.out="TypeTD_tablexport.fa")

write.fasta(as.list(c(EcoliK12$Sequence,subset(BioB_motifs_clean,BioB_motifs_clean$Motif=="TG")$Sequence)), 
            names=c(EcoliK12$Entry,subset(BioB_motifs_clean,BioB_motifs_clean$Motif=="TG")$Entry), 
            as.string=FALSE, file.out="TypeTG_tablexport.fa")

write.fasta(as.list(c(EcoliK12$Sequence,subset(BioB_motifs_clean,is.na(BioB_motifs_clean$Motif))$Sequence)), 
            names=c(EcoliK12$Entry,subset(BioB_motifs_clean,is.na(BioB_motifs_clean$Motif))$Entry), 
            as.string=FALSE, file.out="TypeNA_tablexport.fa")

